### ISSUE TITLE
DefinePlugin with Persistent cache

### ISSUE THREAD

**Issue by vankop** [webpack/webpack#10249]
# Bug report

**What is the current behavior?**

If value provided by DefinePlugin changed when using Persistent cache output will not change.

**If the current behavior is a bug, please provide the steps to reproduce.**

https://github.com/vankop/webpack5-cache-bug

Run `npm run build:dev` than `npm run build:debug` content won't change

**What is the expected behavior?**

Webpack must respect changing values in DefinePlugin (maybe track depended modules and invalidate them)

**Other relevant information:**
webpack version:  5.0.0-beta.11
Node.js version:  10
Operating System:  OSX
Additional tools:

**Comment by sokra** [comment]
see cache.version: https://github.com/webpack/changelog-v5/blob/master/guides/persistent-caching.md#version

**Comment by evilebottnawi** [comment]
@sokra maybe we can throw a warning when a developer uses a cache and `DefinePlugin` without the `version` property?

**Comment by vankop** [comment]
maybe it could be done somehow internally =) I mean define plugin change cache checksum or what ever

**Comment by evilebottnawi** [comment]
I think we should think about it for better 0CJS configuration, because it can be hard to understand and setup for new developers

**Comment by vankop** [comment]
Also I can have a lot definitions and creating such `version` could looks weird

**Comment by evilebottnawi** [comment]
You can use `md4`/`md5`/any hash for creating that, but yes it is looking some hacky

**Comment by sokra** [comment]
yep, I think this could be improved. One issue with using `version` is also that it invalidates the whole cache, while DefinePlugin could invalidate only a single module where the value is used.

**Comment by evilebottnawi** [comment]
bump

**Comment by otakustay** [comment]
From previous issues it is said `DefinePlugin.runtimeValue` can solve this problem?

**Comment by evilebottnawi** [comment]
No, we will fix it

**Comment by otakustay** [comment]
We're relying on an identical cache version to ensure performance between multiple builds with different `DefinePlugin` inputs.

In our build, we first build a stable version with several "feature values" injected via `DefinePlugin`, then have a second insiders build using different "feature values". Since the first build creates a persitent cache, the second build can be very fast.

I'm afraid that using different cache versions can break this and results in a much slower build, so it is better not to have `DefinePlugin` related to `version`.

**Comment by evilebottnawi** [comment]
@otakustay Do you use webpack@5, right?

**Comment by otakustay** [comment]
I'm using webpack@5.0.0-beta.14 currently, we are doing things like:

```js
const baseConfiguration = {
    // entry, devtool, module...
};

const stable = {
    ...baseConfiguration,
    output: {
        filename: '[name]-stable.[chunkhash].js',
    },
    plugins: [
        ...baseConfiguration.plugins,
        new DefinePlugin({
            foo: DefinePlugin.runtimeValue(() => 1),
            bar: DefinePlugin.runtimeValue(() => 1),
        }),
    ],
};

// First build initializes persistent cache
webpack(stable, () => {
    const insiders = {
        ...baseConfiguration,
        output: {
            filename: '[name]-insiders.[chunkhash].js',
        },
        plugins: [
            ...baseConfiguration.plugins,
            // new values
            new DefinePlugin({
                foo: DefinePlugin.runtimeValue(() => 2),
                bar: DefinePlugin.runtimeValue(() => 2),
            }),
        ],
    };

    webpack(insiders, () => {
        console.log('done');
    });
});
```

Comparing to build `stable` and `insiders` together at the same time, this can reduce build speed around 40% in webpack@5, so I'd not like `DefinePlugin` to break cache version and slow down build speed.

**Comment by evilebottnawi** [comment]
Yep, we will fix it

**Comment by evilebottnawi** [comment]
bump

**Comment by kelly-tock** [comment]
glad I found this, had a define value stay the same after first cache. I'm hoping we can provide different define values and still have a speedy production build. used version for now, but basically turns off the caching for my build.

**Comment by kelly-tock** [comment]
also to note various scenarios, locally this can be an issue as well, as we define env variables that are used in webpack to control for example which api server we hit sometimes. I think maybe a workaround is to pass a value in the same manner to the version field for now in the dev configuration to pick up new settings. or as someone mentioned hash the variables combination.

**Comment by alexander-akait** [comment]
> also to note various scenarios, locally this can be an issue as well, as we define env variables that are used in webpack to control for example which api server we hit sometimes.

Do you use webpack-cli? How you define them? I think we can fix it for better DX

**Comment by kelly-tock** [comment]
we just have some flags in our code for example we say `USE_LOCAL_SERVER yarn start`

and in our webpack config we read this in and set specific values in the define plugin. 

for now i'm looping through the values we use and creating a hash to send to version. most of the time locally for dev these don't change, so its not as bad as losing the production build caching speed improvements.

**Comment by alexander-akait** [comment]
Maybe we can allow to set `version`, not only string, but alos as object and do hashing it on own side :thinking:

**Comment by kelly-tock** [comment]
I like the idea mentioned that it could invalidate only things that use define plugin values.

**Comment by sokra** [comment]
fixed by #12896
